src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs

```diff
@@ -401,6 +501,162 @@ private static void AssertNoImplicitInterpolatedStringHandlerConversions(Immutab
             }
         }
 
+        private ImmutableArray<BoundExpression>? TryConvertPartsFromHandlerToConcat(ImmutableArray<BoundExpression> optimizedParts)
+        {
+
+            if (optimizedParts.Length >= 5)
+            {
+                // Prefers handler calls to concat (string[]) for 5 operands or more.
+                return null;
+            }
+            if (!optimizedParts.All(p => p is BoundCall { Method.Name: BoundInterpolatedString.AppendLiteralMethod } or BoundCall { Method.Name: BoundInterpolatedString.AppendFormattedMethod, Arguments: [{ Type.SpecialType: SpecialType.System_String }] }))
+            {
+                return null;
+
+            }
+            var builder = ArrayBuilder<BoundExpression>.GetInstance();
+            foreach (var part in optimizedParts)
+            {
+                var partCall = (BoundCall)part;
+                Debug.Assert(partCall.Method.Name != BoundInterpolatedString.AppendLiteralMethod
+                    || partCall.Arguments[0] is BoundLiteral);
+                builder.Add(
+                    partCall.Method.Name == BoundInterpolatedString.AppendLiteralMethod
+                    ? partCall.Arguments[0]
+                    : new BoundStringInsert(partCall.Syntax, partCall.Arguments[0], null, null, false)
+                );
+            }
+            return builder.ToImmutableAndFree();
+        }
+
+        private (ImmutableArray<BoundExpression> Calls, int IncreasedLiteralLength, int FilledHolesCount) OptimizeAppendFormattedCalls(ImmutableArray<BoundExpression> parts)
+        {
+            var result = ArrayBuilder<BoundExpression>.GetInstance();
+            var currentStringConst = PooledStringBuilder.GetInstance();
+            (BoundCall? AppendLiteralCall, BoundLiteral? StringLiteral) reusable = default;
+            BoundExpression? lastAppendReceiver = null;
+            int filledHolesCount = 0;
+            int increasedLiteralLength = 0;
+
+            BoundExpression createAppendLiteralCallFromBoundLiteral(BoundLiteral literal, BoundExpression receiver) =>
+                _factory.InstanceCall(receiver, BoundInterpolatedString.AppendLiteralMethod, literal);
```

Comment:
#74274 removed this `InstanceCall`. What should I write instead of this?

Reply:
I will retore this definition:

```cs
public BoundExpression InstanceCall(BoundExpression receiver, string name, bool disallowExpandedNonArrayParams, BoundExpression arg0)
    => MakeInvocationExpression(BinderFlags.None, this.Syntax, receiver, name, disallowExpandedNonArrayParams, [arg0], this.Diagnostics);
```

Reply:
https://github.com/dotnet/roslyn/issues/74275 - we can't restore this method.

Reply:
The concern is that we can't do overload resolution at all here. Since we don't need overload resolution to find the specific `AppendLiteral(string)` method, we can just get the method symbol and use it directly.

Reply:
Could you tell me what class and method I should use to generate `AppendLiteral(string)` symbol to deal with the following case?

https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoBBmABM3AYVwG9NcL88EUAGXAWQAoBKU8yzhAdlwBIARCRgQwAUwD2AMyYMWAX2GjJMuYpHjpshSQAiEAC5iAKlHEA6AHISA7uuVa1Szap0aV2+QIDcHCvMx5IA==

```cs
public string M()
    => return $"{nameof(M)}{nameof(M)}{nameof(M)}{DateTime.Now}{nameof(M)}{nameof(M)}{nameof(M)}";
```

I do not know other ways to generate one without registering it to WellKnownMember or SpecialMember.

Reply:
You'd simply look through the handler type and find the `AppendLiteral(string)` method. That's it.

Reply:
I'm investing how to do, but should I `Debug.Assert` that `.Type` of `receiver` (the handler expresion) is not null and that the method exists in the handler type?

Reply:
> You'd simply look through the handler type and find the `AppendLiteral(string)` method. That's it.

The method should be accessible at the use-site too, unless we expect it to be public (and check this).

Reply:
How about this? (local function)

```cs
 MethodSymbol? getAppendLiteralCallFromReceiver(BoundExpression receiver)
 {
     Debug.Assert(receiver.Type is not null);
     foreach (var candidate in receiver.Type!.GetMembers(BoundInterpolatedString.AppendLiteralMethod))
     {
         if (candidate is not MethodSymbol methodCandidate) continue;
         if (methodCandidate.IsStatic) continue;
         if (methodCandidate is not { Parameters: [{ Type.SpecialType: SpecialType.System_String }] }) continue;
         if (methodCandidate.DeclaredAccessibility != Accessibility.Public) continue;
         return methodCandidate;

     }
     return null;
 }
```

Reply:
I'd probably condense all of that into a single pattern, but otherwise it looks about right.

Reply:
Do I have to aggregate some conditional expressions, not only chain them with `or` and `||`?

Reply:
I don't really know what that question means?

Reply:
I mean whether you prefer the following

```cs
         if (
             candidate is not MethodSymbol { Parameters: [{ Type.SpecialType: SpecialType.System_String }], DeclaredAccessibility: Accessibility.Public} methodCandidate
                 || methodCandidate.IsStatic
          )
          {
              continue;
          }
         return methodCandidate;
```

to the follwing:

```cs
         if (
             candidate is not MethodSymbol methodCandidate
                 || methodCandidate.IsStatic
                 || methodCandidate is not { Parameters: [{ Type.SpecialType: SpecialType.System_String }] }
                 || methodCandidate.DeclaredAccessibility != Accessibility.Public
          )
          {
              continue;
          }
         return methodCandidate;
```

---


NOTE: we can't write the following due to https://github.com/dotnet/csharplang/discussions/8323:
```cs
         if (
             candidate is not MethodSymbol { Parameters: [{ Type.SpecialType: SpecialType.System_String }], DeclaredAccessibility: Accessibility.Public} methodCandidate
                 or { IsStatic: false }
          )
          {
              continue;
          }
         return methodCandidate;
```

The following is OK:

```cs
         if (
             candidate is not (
                     MethodSymbol { Parameters: [{ Type.SpecialType: SpecialType.System_String }], DeclaredAccessibility: Accessibility.Public} methodCandidate
                         and { IsStatic: false }
                 )
          )
          {
              continue;
          }
         return methodCandidate;
```

Reply:
None of the above. I mean
```
if (candidate is MethodSymbol { IsStatic: false, Parameters: [{ Type.SpecialType: SpecialType.System_String]}, DeclaredAssessibility: Assessibility.Public } methodCandidate)
{
    return methodCandidate;
}
```